src/impl_/pyclass/lazy_type_object.rs

```diff
@@ -27,7 +33,10 @@ struct LazyTypeObjectInner {
     value: GILOnceCell<PyClassTypeObject>,
     // Threads which have begun initialization of the `tp_dict`. Used for
     // reentrant initialization detection.
+    #[cfg(not(Py_GIL_DISABLED))]
     initializing_threads: GILProtected<RefCell<Vec<ThreadId>>>,
+    #[cfg(Py_GIL_DISABLED)]
+    initializing_threads: Mutex<RefCell<Vec<ThreadId>>>,
```

Comment:
I think you don't need the `RefCell` any more and can drop that layer:

```suggestion
    initializing_threads: Mutex<Vec<ThreadId>>,
```
Also, given that the locking is extremely short lived and doesn't execute arbitrary python code, I think we can just use this form on all builds to keep it simple. There shouldn't be risk of deadlock with the GIL as far as I can tell.